modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    pragma "no doc"
+    inline proc _withinBounds(idx: int): bool {
+      return (idx >= 1 && idx <= _size);
+    }
+
+    //
+    // This call assumes that a lock (acquired by `_enter`) is already
+    // held (if parSafe==true), and releases it before throwing an error if
+    // a bounds check fails.
+    //
+    pragma "no doc"
+    inline proc _boundsCheckLeaveOnThrow(i: int) throws {
+      if !_withinBounds(i) then {
+        _leave();
+        const msg = "List index out of bounds: " + i:string;
+        throw new owned
+          IllegalArgumentError(msg);
+      }
+    }
+
+    pragma "no doc"
+    proc _makeBlockArray(size: int) {
+      type UnmanagedBlock = unmanaged ListBlock;
+      var result = new unmanaged
+        ListBlock(unmanaged ListBlock(eltType), size);
+      return result;
+    }
+
+    //
+    // A call to this assumes that if no free blocks remain, then all free
+    // blocks have been filled and we need to resize the blocks array.
+    //
+    pragma "no doc"
+    proc _maybeResizeBlockArray() {
+
+      const lastBlockIdx = _getBlockIdx(_size);
+      if lastBlockIdx < _blocks.capacity then
+        return;
+
+      var _nblocks = _makeBlockArray(_blocks.capacity * 2);
+      for i in 1.._blocks.capacity do
+        _nblocks[i] = _blocks[i];
+      delete _blocks;
+      _blocks = _nblocks;
+    }
+
+    pragma "no doc"
+    proc _maybeAcquireMem(amount: int) {
+      //
+      // Should only be in this state after an init or a clear.
+      //
+      if _blocks == nil then {
+        _sanity(_totalCapacity == 0);
+        _sanity(_size == 0);
+        _blocks = _makeBlockArray(_initialBlockCapacity);
+        _blocks[1] = new unmanaged ListBlock(eltType, _initialCapacity);
+        _totalCapacity += _initialCapacity;
+      }
+
+      const remaining = _totalCapacity - _size;
+      if remaining > amount then
+        return;
+
+      var lastBlockIdx = _getBlockIdx(_size);
+      var req = amount - remaining;
+
+      while req > 0 do {
+        _maybeResizeBlockArray();
+
+        ref oblock = _blocks[lastBlockIdx];
+        lastBlockIdx += 1;
+        ref nblock = _blocks[lastBlockIdx];
+        _sanity(nblock == nil);
+        _sanity(oblock != nil);
+
+        nblock = new unmanaged ListBlock(eltType, oblock.capacity * 2);
+        _totalCapacity += nblock.capacity;
+        req -= nblock.capacity;
+      }
+    }
+
+    pragma "no doc"
+    proc _maybeReleaseMem(in amount: int) {
+      if _blocks == nil || _totalCapacity == 0 then
+        return;
+
+      //
+      // TODO: Worry about writing this _after_ we get the test functions to
+      // work once again.
+      //
+      
+    }
+
+    //
+    // Shift elements including and after index one to the right in memory,
+    // possibly resizing.
```

Comment:
Add "Assumes the element at idx contains uninitialized memory (i.e. it was already deinited if that is necessary)"

Reply:
Could you explain what you mean? The element at `idx` has already been initialized by some other operation. After the shift the memory at `idx` is in an invalid state, are you saying I should call the destructor on it?

Reply:
Oh sorry, I had the wording wrong. The comment should just say that this function will leave `idx` containing uninitialized memory.

Do you call the destructor on it elsewhere? I thought you did. It's just a matter of commenting this function's assumptions, so it could say something like "the caller should call the destructor on it if necessary"

Reply:
Expand doesn't call a destructor, because it expects the new element to be inserted into `idx` after the move.

Collapse expects a destructor call to have occurred before the elements are shifted left (else, that element is lost to oblivion and the destructor can never be called).

I will write these down.